# _Python for Scientific Data Analysis_


# Basic Python
## Section 1: Printing, Running Code, and Variables

### Hello World

The first step in learning Python or any other language is to ask a piece of code to print a statement.   

 To do this from command line, first you type ``python`` in your terminal window, which will bring up the Python prompt. 
 
 Print statements with Python 3 are enclosed by parentheses. 

E.g.

```print("hello world")```

 will print ...
 

In [1]:
print("hello world")
print("hello world2 or 3 or 4")
#print(hello world)

hello world
hello world2 or 3 or 4


Note: if you do not enclose this string with ", then it will trigger a traceback error because _hello world_ is not a named variable.  If you try to treat _hello world_ as a variable you will get a traceback error: e.g. ``` >>> hello world = 5``` triggers an error.   However, the following does not trigger an error:

In [2]:
hello_world = 7
hello_world='yo'
hello,world = 7,11
#hello_world=7,11  #print('try this\n',hello_world,hello_world[0],hello_world[1])
hello_world = 'yo','checkit'

print(hello_world)
#print(hello,world)
print(hello)
print(world)

('yo', 'checkit')
7
11


where then you type
```print(hello_world)``` or ```print(hello,world)``` at the prompt.

What happened in these four examples?   In the first two, we declared a single variable, ```hello_world```, as an integer 7 or a string `yo`.   But Python allows you to declare multiple variables in a single line.   This is shown in third and fourth example: here, ```hello``` and ```world``` are two separate variables simultaneously set to two different integers and two different strings, respectively.  Cool, right?   



### Running code ###

There are a couple of ways to run Python code

#### From Terminal/Powershell

Here you have a couple of options.  The key unit of Python code is a script.  They have a suffix of .py (e.g. ``testme.py``).  A script is just a series of instructions (do this, do that, etc.).  A script may contain many Python ``functions`` (more on that later), which allow you to execute non-trivial tasks.

To run a full Python script from Terminal/Powershell it is pretty easy:

```
python [nameofscript]
```

If I have a script and want a lot of the details hardwired, I do this (exceptions being argument passing: we will deal with that later).

An example: look at ex1.py.  

``python ./code_ipynb/ex1.py``

#prints out

hello world

yo

5

yo checkit

7 11

#### From Python Prompt

When not working with Jupyter Notebooks, this is my standard way of executing Python code.

To access it, simply call ``python``

e.g.
```
(base) thaynecurrie@RR189794IP01 code_ipynb % python

Python 3.12.4 | packaged by Anaconda, Inc. | (main, Jun 18 2024, 10:07:17) [Clang 14.0.6 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> 
```

Now, this will not let you do ``python ex1.py`` (you 'can' execute scripts within Python prompt but it looks messy). 

But you can do something different later that is quite powerful: access executable functions within scripts (e.g. "from the script, _import_ something very specific, then execute that specific thing, usually in a tuned way").  For now, just keep this in the back of your mind.

#### From Within Jupyter Notebooks

You can execute scripts just like you can do from Terminal/Powershell by adding a ``!`` within Notebooks. This will literally say "hey, run this script in a shell (e.g. as if you are in Terminal/Powershell)".

E.g.

In [3]:
!python ./code_ipynb/ex1.py #from notebooks

hello world
yo
5
yo checkit
7 11


Another way of executing scripts is with the ``%run`` command: e.g. ``%run myscript.py``.  The difference here is that if a variable is named and given a value in that script, then you can access it after the run command

In [7]:
#now just type again to run again


%run ./code_ipynb/ex1.py

hello world
yo
5
yo checkit
7 11


In [8]:
print(hello) #will return 7
print(hello+5) #will return 12
print(hello*world) #will return 77

7
12
77


In [9]:
#watch this ... we will cover this later

from code_ipynb.ex1b import dosomething

dosomething()

hello world
yo
5
yo checkit
7 11


Note: the above -- importing a function and watching it do something automatically -- is just for demonstration purposes.  You should not actually run code this exact way (we will cover the varieties of code execution below).

### Variables and Printing
Okay, now we are going to make things slightly more complicated.   When you print something with Python (or any other language), you want to control the format of what you write: do you want to write a string? An integer?  decimal? floating point?  That's where _formatted_ print comes into play.

#### Formatted Printing: The standard way

Python's _formatted_ print statements come after a ```%``` for a print statement: e.g. ```print("my name is %[formatting statement]" % [variable] ```).   

Fortunately, these are pretty easy to figure out: ```%s``` (e.g. "yo") for string, ```%d``` or ```%i``` for decimal integers (e.g. 5), ```%f``` for floating point (e.g. 3.141592), ```%e``` for lower case exponential notation (e.g. 1.41569e+03), etc.  There are a few others but these are the most important ones for now.   

Now, look at ex5a.py.  It should read as follows

```sh
my_name = 'The Great Cornholio'
my_age = 21 # ya right
my_height = 72 # with stilts
my_weight = 9 # lbs
my_jump = 15 #feet
my_eyes = 'White'

print("My name is %s." % my_name)
print("I am %d inches tall and I weigh %d pounds." % (my_height,my_weight))
print("Even though my eyes are %s, I can jump %d feet in the air" % (my_eyes,my_jump))
print("If you add %d, %d, and %d together you get %d.  Scared now of the %s?" % (my_age,my_height,my_jump,my_age+my_height+my_jump,my_name))

```


If you execute this code, you should see:

```
My name is The Great Cornholio.
I am 72 inches tall and I weigh 9 pounds.
Even though my eyes are White, I can jump 15 feet in the air
If you add 21, 72, and 15 together you get 108.  Scared now of the The Great Cornholio?

```

Notice another thing.  In the source code we put #s all over the place.  These look like comments, right?  That they are! These do not print out when you execute the code.

#### Formatted Print Statements: The second, more 'modern' way

Another way to do formatted print statements are with curly brackets ```{ }``` and a format statement.  Here's an example

In [10]:
statement='far out'
print("this statement is {}, dude".format(statement))


print("this statement is {}, dude, {}".format(statement,statement))

this statement is far out, dude
this statement is far out, dude, far out


You should see ``this statement is far out, dude``.

Now, slightly more complicated

In [11]:
print("this statement is {0} {1}, dude".format(statement,'seriously'))

print("this statement is {0} {1}, dude".format(statement,statement + ' seriously'))

this statement is far out seriously, dude
this statement is far out far out seriously, dude


See what we've done here? we now have two variables with formatted print.  

#### More control over formatting
Now we will take the second formatting approach and make this more complicated ...

In [12]:
print("this statement is {0} {1}, dude".format(statement,'seriously'))

print("this statement is {0:.1s} {1:s}, dude".format(statement,'seriously'))

print("this statement is {0:.1s} {1:.2s}, dude".format(statement,'seriously'))


this statement is far out seriously, dude
this statement is f seriously, dude
this statement is f se, dude


which prints out ``this statement is f seriously, dude`` and ``this statement is f se, dude ``.  See what happened? if not, look to see what happens here:

In [13]:

import numpy as np #don't worry about what this line means right now. just think "command allowing us to do stuff"
evals=np.e
print("this stupid number is {0:f}, not {1:.3f} or {2:.5f}, \
but if we want more than the \
 default precision of {3:d} spaces after the decimal, \
 we can write it as something different like {4:.10f}".format(evals,evals,evals,6,evals))


this stupid number is 2.718282, not 2.718 or 2.71828, but if we want more than the  default precision of 6 spaces after the decimal,  we can write it as something different like 2.7182818285


which prints out as ``this stupid number is 2.718282, not 2.718 or 2.71828,``
`` but if we want more than the  default precision of 6 spaces after the decimal,``  
``we can write it as something different like 2.7182818285`` 

Note: all of this is on one line: the `\` signifies "continue the line".

So what happened there? we can control the output.  For example, in the first case, we said "format this as a floating point".  In the second "format this as a floating point but only leave 3 spaces after the decimal".  And so on.  By default, Python 3.x prints out 6 decimal spaces for a floating point.   If you want more precision, then you can just add more space, as we do in the last formatted print for the variable `evals`.

Now what if we wanted to use `%` for formatted print with all that control?.  Here you go:

In [14]:

print("this stupid number is %f, not %.3f or %.5f, \
but if we want more than the \
 default precision of %d spaces after the decimal, \
 we can write it as something different like %.10f" %(evals,evals,evals,6,evals))


this stupid number is 2.718282, not 2.718 or 2.71828, but if we want more than the  default precision of 6 spaces after the decimal,  we can write it as something different like 2.7182818285


## A Note on Using Code: Functions, Import and \_Main_

Earlier, we discussed how to execute python scripts containing a series of instructions/tasks.  E.g. ``python [name of python script].py``
The script ``ex1.py`` is such an example.  



```
python ex1.py #from command line

#prints out 
#hello world
#yo
#5
#yo checkit
#7 11
```

But this is not the only way to use code in scripts.  We can actually pull pieces of code from a given script and reuse them for different purposes.

E.g. ``ex2.py``

```
python ./code_ipynb/ex2.py

#prints out
#[0.03232323 0.04266667 0.05637172]
```

Now, in interactive mode -- Here you have a few options.  i) You can write code line by line (i.e. interactively). If you want to do something non-trivial, you will likely need to ``import`` a python ``function`` from a ``module`` to do this.  This works so long as the piece of code is in your Python path (either local or within your path variable). ii) You can also import and run entire functions to do calculations.   The latter is (IMO) the standard way to execute Python code.

   In either case, to get started, you do the following


from Terminal, you first type ``python``.  "Running" python interactively (i.e. as in option i) )then has the following structure (note, each line is a separate line in the interactive mode ... i.e. a separate ">>>")


An example ...

```
>>> import numpy as np #we will encounter this "numpy" library a lot!
>>> d_tel = 8.2 #telescope diameter (meters)
>>> wvlh_microns=[1.25,1.6,2.18] #wavelength (microns)
>>> fwhm=1.028*206265*np.array(wvlh_microns)*1e-6/d_tel
>>> print(fwhm)

#this will print out
#[0.03232323 0.04137374 0.05637172]
```

_You can do the same thing within a Jupyter notebook_

In [15]:
import numpy as np
d_tel=8.2 #telescope diameter (meters)
wvlh_microns=[1.25,1.6,2.18] #wavelength (microns)
fwhm=1.028*206265*np.array(wvlh_microns)*1e-6/d_tel
print(fwhm)

[0.03232323 0.04137374 0.05637172]


Writing things out by hand is tedious.  So we can store this the calculation in a Python script and call the script like we have done before (early in this notebook).  

This is reproduced as ``ex2.py`` in the code_ipynb subdirectory.  The source code is below

```
import numpy as np

def calc_fwhm(d_tel=8.2,wvlh_microns=[1.25,1.65,2.18]):

 fwhm=1.028*206265*np.array(wvlh_microns)*1e-6/d_tel

 print(fwhm)
 return fwhm


if __name__ == '__main__':

 calc_fwhm()

```

An easy way to execute this code is then ``python ./code_ipynb/ex2.py`` from this notebook directory (or, equivalently, ``python ex2.py`` if you are within the subdirectory containing ex2.py). 

Or, if we are using this notebook, you create a cell and do an equivalent calculation

In [16]:
!python ./code_ipynb/ex2.py

the fwhm is  [0.03232323 0.04266667 0.05637172]


But what if we want to re-use the code with slightly different variations on the calculations?  Or if we want the option to just call the script from Terminal/Powershell/Jupyter notebooks _*or*_ re-use the code with slightly different numbers for its input?

That is the _*function*_ syntax (denoted as ``def``) _and_ the last bit of code with ``if __name__ == '__main__':`` come in.  

With ``if __name__ == '__main__':``, we can just run the entire thing as a script by typing ``python [path_to_]ex2.py`` just like before.  
With ``def``, we can import from this script to re-use the same type of calculation we've been doing but with slightly different numbers.  To do this we _import_ the function that does this calculation from the python script as follows. 

```
(base) thaynecurrie@RR189794IP01 Basic % python ./code_ipynb/ex2.py
```

```
>>>from ex2 import calc_fwhm
>>>calc_fwhm()
#will print out [0.03232323 0.04266667 0.05637172]

>>>calc_fwhm(d_tel=30)
#prints out [0.00883502 0.01166222 0.01540827]
```

you can easily see how this is more powerful.

In [17]:
#or in a Jupyter Notebook cell

from code_ipynb.ex2 import calc_fwhm 
#note: we have to type from code_ipynb.ex2, 
#  not from ex2, because this Notebook is a directory above code_ipynb

calc_fwhm()
#prints out [0.03232323 0.04266667 0.05637172]
calc_fwhm(wvlh_microns=[0.8,3.78,10]) 
#prints out [0.02068687 0.09774546 0.25858588]

result=calc_fwhm(d_tel=30,wvlh_microns=[0.8,3.78,10]) 

print(result)
type(result)

the fwhm is  [0.03232323 0.04266667 0.05637172]
the fwhm is  [0.02068687 0.09774546 0.25858588]
the fwhm is  [0.00565441 0.02671709 0.07068014]
[0.00565441 0.02671709 0.07068014]


numpy.ndarray

And now, we can combine the function within other code to present nicely formatted output.

In [18]:
import numpy as np

from code_ipynb.ex2 import calc_fwhm

#d_tel = 8.2 #telescope diameter (meters), the default for ex2
lam_tel= np.array([1.25,1.65,2.18]) #the default for ex2
     #wavelength (microns) note: we don't call it lambda because that is a separate, canned Python thing

#d_tel=np.array([8.2,30,50])

d_tel=8.2

result=calc_fwhm(d_tel=d_tel)
#print((result[0]))


for i in range(0,len(lam_tel)):

 print("The telescope (diffraction-limited) FWHM in arcseconds given a diameter of {0:.1f} meters \
 for wavelength of {1:.4f} is {2:.4f}".format(d_tel,lam_tel[i],result[i]))

the fwhm is  [0.03232323 0.04266667 0.05637172]
The telescope (diffraction-limited) FWHM in arcseconds given a diameter of 8.2 meters  for wavelength of 1.2500 is 0.0323
The telescope (diffraction-limited) FWHM in arcseconds given a diameter of 8.2 meters  for wavelength of 1.6500 is 0.0427
The telescope (diffraction-limited) FWHM in arcseconds given a diameter of 8.2 meters  for wavelength of 2.1800 is 0.0564


We will discuss Python functions and import in more detail later. We will also add a way to re-use Python scripts from Terminal/Powershell/Notebooks with different variables later.  

My preferred method for submitting homework will be with Jupyter notebooks. 


 
 For more details, see this nice discussion of the Python interpreter on the official site, here: <https://docs.python.org/3/tutorial/interpreter.html>

### Python PATH for importing functions, installing programs

What if you are working in some directory and want to import and then use a Python function?  It is usually good practice to keep all your code -- at least detailed packages --  in a place well separated from what you are working on.  

In order for you to import and utilize that code, it has to be in your Python _path_ .  How do you know what your path is?  Here's how:

In [19]:
#from os import sys #this may also work
import sys
print(sys.path)

print('')
#or, more elegantly & better formatted...
print('2nd version')
print('')
print('\n'.join(sys.path))
print('')
print('3rd version: just PythonPATH')
#stuff that you added to sys.path
import os
pypath=os.environ['PYTHONPATH'].split(os.pathsep)
print('\n'.join(pypath))

['/Users/thaynecurrie/Research/coding/Python/ScientificPythonNotes_2026/Basic', '/Users/thaynecurrie/Research/Github/repositories/charis-dpp_test/python', '/Users/thaynecurrie/Research/Data/stellar/Baffles/BAFFLES', '/Users/thaynecurrie/anaconda3/lib/python312.zip', '/Users/thaynecurrie/anaconda3/lib/python3.12', '/Users/thaynecurrie/anaconda3/lib/python3.12/lib-dynload', '', '/Users/thaynecurrie/anaconda3/lib/python3.12/site-packages', '/Users/thaynecurrie/anaconda3/lib/python3.12/site-packages/aeosa', '/Users/thaynecurrie/Research/Github/repositories/orvara_py311/orvara']

2nd version

/Users/thaynecurrie/Research/coding/Python/ScientificPythonNotes_2026/Basic
/Users/thaynecurrie/Research/Github/repositories/charis-dpp_test/python
/Users/thaynecurrie/Research/Data/stellar/Baffles/BAFFLES
/Users/thaynecurrie/anaconda3/lib/python312.zip
/Users/thaynecurrie/anaconda3/lib/python3.12
/Users/thaynecurrie/anaconda3/lib/python3.12/lib-dynload

/Users/thaynecurrie/anaconda3/lib/python3.12/sit

The ``sys.path`` variable prints out your entire list of paths available for Python to import functions.   The `` '\n'.join(sys.path`` shows the same information with each path on a new line.

The last print statement shows paths that are in your ``$PYTHONPATH`` environment variable.  This is something that you have to set in your shell.  See my example in class ...

Now, if you HAVEN'T set your PYTHONPATH variable in a shell startup script anywhere (on a newer Mac, this is usually .zshrc; on newer Linux this is probably .bashrc) the last command will likely throw an error.

If you don't know whether you have set a PYTHONPATH, you can find out by the following

In [20]:
pypath=os.environ.get('PYTHONPATH','')
pypath

'/Users/thaynecurrie/Research/Github/repositories/charis-dpp_test/python/:/Users/thaynecurrie/Research/Data/stellar/Baffles/BAFFLES/'

Now what if there's a Python package that you want to install?  Well, you have a couple of options.  The two most common installation methods are using ``conda`` -- if you have the Anaconda distribution of Python -- or ``pip``.   If the thing you are wanting to install is a very nice, official package (e.g. ``numpy``), it should have its own installer script that you can use from ``conda`` or ``pip``.  

Examples

(installing or upgrading ``numpy``) (note: you should only need to upgrade numpy to new versions: an installation comes standard with Anaconda)

```
conda install numpy
```

or

```
pip install numpy
```

Some caveats.  

- First, ``conda`` is SLOW: it will carefully check for (what it thinks are) package incompatibilities.  And if you are not careful, it may start overwriting packages you currently have to solve the incompatibility.

- Or conversely, ``pip`` might do an installation or upgrade that breaks packages a bit more easily.

I've dealt with both cases.

Good policies for installing new packages.  If you are unsure what will happen and do not want to spend time debugging an installation where conda/pip break things you care about in order to install some other package, try the following

- do a "dry run" with pip. This will simulate a package download and what your computer will do in order to install it.  If you see "xyz package [thing you rely on] will be updated", this is your time to assess whether an update might break your environment.

  ``pip install <package_name> --dry-run``

- Make a hard backup of your entire anaconda path and backup your computer before major code installations

- Have a base conda environment for your core code base.  If you are doing something unique (e.g. this class) I would recommend that you create a _separate_ conda environment.  E.g.
  
  ``conda create -n myenv``

  
